syntax = "proto3";

message ConsentPayload {

    enum PayloadType {
        TYPE_UNSET = 0;
        PAYLOAD_ADD_CLIENT = 1;
        PAYLOAD_APPROVE_CONSENT = 2;
        PAYLOAD_REQUEST_CONSENT = 3;
        PAYLOAD_DECLINE_CONSENT = 4;
        PAYLOAD_REVOKE_CONSENT = 5;
    }

    PayloadType payload_type = 1;

    Client create_client = 2;
    ActionOnAccess approve_consent = 3;
    ActionOnAccess request_consent = 4;
    ActionOnAccess decline_consent = 5;
    ActionOnAccess revoke_consent = 6;
}

message ActionOnAccess {

    enum ActionType {
        TYPE_UNSET = 0;
        REQUESTED = 1;
        APPROVED = 2;
        DECLINED = 3;
    }

    string dest_pkey = 1;
    string src_pkey = 2;
    ActionType action_type = 3;
}

message Client {
    string public_key = 1;
    repeated Permission permissions = 2;
}

message Permission {
    enum PermissionType {
      PERMISSION_TYPE_UNSET = 0;

      GET_CONSUMERS_LIST = 100;
      ADD_DATA = 101;
      READ_OWN_DATA = 102;
      READ_DATA = 103;
      GET_REQUESTS_LIST = 104;
      APPROVE_CONSENT = 105;
      DECLINE_CONSENT = 106;
      REVOKE_CONSENT = 107;

      GET_ACADEMICS_LIST = 200;
      REQUEST_CONSENT = 201;

      READ_DATA_TRACE = 300;
      READ_OWN_DATA_TRACE = 301;

    }

    PermissionType type = 1;
}
